controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -451,38 +571,130 @@ func (pp *portPublisher) updateEndpoints(endpoints *corev1.Endpoints) {
 			}
 		}
 	}
-	pp.exists = true
 	pp.addresses = newAddressSet
+}
 
-	pp.metrics.incUpdates()
-	pp.metrics.setPods(len(pp.addresses.Addresses))
-	pp.metrics.setExists(true)
+func (pp *portPublisher) updateEndpointSlicesSet(endpointslice *dv1beta1.EndpointSlice) {
+	newAddressSet := pp.endpointSliceToAddresses(endpointslice)
+	if len(newAddressSet.Addresses) == 0 {
+		for _, listener := range pp.listeners {
+			listener.NoEndpoints(true)
+		}
+	} else {
+		add, remove := diffAddresses(pp.addresses, newAddressSet)
```

Comment:
It looks like when we see an EndpointSlice add or update, we are treating that EndpointSlice as the entire set of addresses when actually there could be other EndpointSlice objects which exist for that service.

I think what we need to do is slightly more complex:
* For EndpointSlice adds, we can update `pp.addresses` by adding the addresses in the EndpointSlice
* For updates, we probably want to start by removing all the addresses in the old EndpointSlice from `pp.addresses` and then adding the ones in the new EndpointsSlice

In either case, we'd want to do a diff of the previous `pp.addresses` to the current `pp.addresses` to determine which updates to send to the listener.

Reply:
+1 on that I think one of the main benefits of slices is that you can have multiple objects for a service. So this is certainly a valid case.

Reply:
This answers a few questions I had! Thank you both for your comments, I think this has been the trickiest part of adding support for this feature so far. Will change things around based on the feedback 👍